### PR TITLE
Fix #4212 Error while indexing (RsExternCrateReexportIndex)

### LINKED ISSUE #4212: Error while indexing (RsExternCrateReexportIndex)
```
Error while indexing ~/workspace/rust/src/test/ui/unnecessary-extern-crate.rs
To reindex this file IDEA has to be restarted

com.intellij.openapi.project.IndexNotReadyException: Please change caller according to com.intellij.openapi.project.IndexNotReadyException documentation
	at com.intellij.openapi.project.IndexNotReadyException.create(IndexNotReadyException.java:81)
	at com.intellij.util.indexing.FileBasedIndexImpl.handleDumbMode(FileBasedIndexImpl.java:770)
	at com.intellij.util.indexing.FileBasedIndexImpl.ensureUpToDate(FileBasedIndexImpl.java:719)
	at com.intellij.util.indexing.FileBasedIndexImpl.ensureUpToDate(FileBasedIndexImpl.java:702)
	at com.intellij.psi.stubs.StubIndexImpl.doProcessStubs(StubIndexImpl.java:306)
	at com.intellij.psi.stubs.StubIndexImpl.processElements(StubIndexImpl.java:282)
	at com.intellij.psi.stubs.StubIndex.processElements(StubIndex.java:65)
	at org.rust.lang.core.stubs.index.RsIncludeMacroIndex$Companion.getIncludingModInternal(RsIncludeMacroIndex.kt:56)
	at org.rust.lang.core.stubs.index.RsIncludeMacroIndex$Companion.access$getIncludingModInternal(RsIncludeMacroIndex.kt:29)
	at org.rust.lang.core.stubs.index.RsIncludeMacroIndex$Companion$getIncludingMod$1.compute(RsIncludeMacroIndex.kt:45)
	at com.intellij.psi.util.CachedValuesManager.lambda$getCachedValue$0(CachedValuesManager.java:148)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:54)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$1(CachedValueBase.java:240)
	at com.intellij.openapi.util.RecursionManager$1.doPreventingRecursion(RecursionManager.java:113)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:71)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:241)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:43)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValueFromExHolder(CachedValuesManagerImpl.java:72)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:45)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:147)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:129)
	at org.rust.lang.core.stubs.index.RsIncludeMacroIndex$Companion.getIncludingMod(RsIncludeMacroIndex.kt:43)
	at org.rust.lang.core.macros.RsExpandedElement$Companion.getContextImpl(RsExpandedElement.kt:31)
	at org.rust.lang.core.psi.ext.RsExternCrateItemImplMixin.getContext(RsExternCrateItem.kt:42)
	at com.intellij.psi.util.PsiTreeUtil.getContextOfType(PsiTreeUtil.java:641)
	at com.intellij.psi.util.PsiTreeUtil.getContextOfType(PsiTreeUtil.java:620)
	at org.rust.lang.core.psi.ext.PsiElementKt.getContextualFile(PsiElement.kt:269)
	at org.rust.lang.core.psi.ext.RsElementKt.getCargoProject(RsElement.kt:42)
	at org.rust.lang.core.resolve.ImplLookup$Companion.relativeTo(ImplLookup.kt:792)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImplKt.resolvePath$default(RsPathReferenceImpl.kt:83)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:72)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl$Resolver.invoke(RsPathReferenceImpl.kt:70)
	at org.rust.lang.core.resolve.ref.RsResolveCache$resolveWithCaching$$inlined$run$lambda$1.invoke(RsResolveCache.kt:98)
	at org.rust.lang.core.resolve.ref.RecursionGuardWrapper$sam$com_intellij_openapi_util_Computable$0.compute(RecursionGuardWrapper.kt)
	at com.intellij.openapi.util.RecursionManager$1.doPreventingRecursion(RecursionManager.java:113)
	at org.rust.lang.core.resolve.ref.RecursionGuardWrapper.doPreventingRecursion(RecursionGuardWrapper.kt:15)
	at org.rust.lang.core.resolve.ref.RsResolveCache.resolveWithCaching(RsResolveCache.kt:98)
	at org.rust.lang.core.resolve.ref.RsResolveCache.resolveWithCaching(RsResolveCache.kt:88)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedCachedMultiResolve(RsPathReferenceImpl.kt:63)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.advancedMultiResolve(RsPathReferenceImpl.kt:59)
	at org.rust.lang.core.resolve.ref.RsPathReferenceImpl.multiResolve(RsPathReferenceImpl.kt:51)
	at com.intellij.psi.PsiPolyVariantReferenceBase.resolve(PsiPolyVariantReferenceBase.java:47)
	at org.rust.lang.core.resolve.ref.RsReferenceBase.resolve(RsReferenceBase.kt:28)
	at org.rust.lang.core.psi.ext.RsVisibilityKt.getVisibility(RsVisibility.kt:90)
	at org.rust.lang.core.psi.ext.RsVisibilityOwner.getVisibility(RsVisibility.kt:21)
	at org.rust.lang.core.psi.ext.RsVisibilityKt.isPublic(RsVisibility.kt:24)
	at org.rust.lang.core.stubs.index.RsExternCrateReexportIndex$Companion.index(RsExternCrateReexportIndex.kt:34)
	at org.rust.lang.core.stubs.StubIndexingKt.indexExternCrate(StubIndexing.kt:21)
	at org.rust.lang.core.stubs.RsExternCrateItemStub$Type.indexStub(StubImplementations.kt:219)
	at org.rust.lang.core.stubs.RsExternCrateItemStub$Type.indexStub(StubImplementations.kt:201)
	at com.intellij.psi.stubs.ObjectStubTree.indexStubTree(ObjectStubTree.java:69)
	at com.intellij.psi.stubs.SerializedStubTree.indexTree(SerializedStubTree.java:107)
	at com.intellij.psi.stubs.StubUpdatingIndex$2.lambda$computeValue$0(StubUpdatingIndex.java:163)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:931)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:57)
	at com.intellij.psi.stubs.StubUpdatingIndex$2.computeValue(StubUpdatingIndex.java:112)
	at com.intellij.psi.stubs.StubUpdatingIndex$2.computeValue(StubUpdatingIndex.java:108)
	at com.intellij.util.indexing.SingleEntryIndexer.map(SingleEntryIndexer.java:40)
	at com.intellij.util.indexing.SingleEntryIndexer.map(SingleEntryIndexer.java:30)
	at com.intellij.util.indexing.impl.MapReduceIndex.mapInput(MapReduceIndex.java:313)
	at com.intellij.util.indexing.VfsAwareMapReduceIndex.mapInput(VfsAwareMapReduceIndex.java:108)
	at com.intellij.util.indexing.impl.MapReduceIndex.calculateUpdateData(MapReduceIndex.java:269)
	at com.intellij.util.indexing.impl.MapReduceIndex.update(MapReduceIndex.java:251)
	at com.intellij.util.indexing.FileBasedIndexImpl.updateSingleIndex(FileBasedIndexImpl.java:1673)
	at com.intellij.util.indexing.FileBasedIndexImpl.lambda$doIndexFileContent$28(FileBasedIndexImpl.java:1603)
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.freezeFileTypeTemporarilyIn(FileTypeManagerImpl.java:634)
	at com.intellij.util.indexing.FileBasedIndexImpl.doIndexFileContent(FileBasedIndexImpl.java:1562)
	at com.intellij.util.indexing.FileBasedIndexImpl.indexFileContent(FileBasedIndexImpl.java:1549)
	at com.intellij.util.indexing.UnindexedFilesUpdater.lambda$indexFiles$0(UnindexedFilesUpdater.java:109)
	at com.intellij.openapi.project.CacheUpdateRunner$MyRunnable.lambda$run$0(CacheUpdateRunner.java:219)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1105)
	at com.intellij.openapi.project.CacheUpdateRunner$MyRunnable.lambda$run$1(CacheUpdateRunner.java:235)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:169)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:537)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:156)
	at com.intellij.openapi.project.CacheUpdateRunner$MyRunnable.run(CacheUpdateRunner.java:232)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:294)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #4212

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/stubs/index/RsExternCrateReexportIndex.kt [hunk 1]
```diff
@@ -31,7 +30,8 @@ class RsExternCrateReexportIndex : StringStubIndexExtension<RsExternCrateItem>()
 
         fun index(stub: RsExternCrateItemStub, sink: IndexSink) {
             val externCrateItem = stub.psi
-            if (!externCrateItem.isPublic) return
+            val isPublic = externCrateItem?.vis != null
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/stubs/index/RsExternCrateReexportIndex.kt]: Let's be more precise here and check that `visRestriction` is also null here

File: src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt [hunk 2]
```diff
@@ -36,7 +36,7 @@ class RsFileStub : PsiFileStubImpl<RsFile> {
 
     object Type : IStubFileElementType<RsFileStub>(RsLanguage) {
         // Bump this number if Stub structure changes
-        override fun getStubVersion(): Int = 176
+        override fun getStubVersion(): Int = 177
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]:
```suggestion
        override fun getStubVersion(): Int = 178
```
Due to #4194

File: src/main/kotlin/org/rust/lang/core/stubs/index/RsExternCrateReexportIndex.kt (hunk 1 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/stubs/index/RsExternCrateReexportIndex.kt]: This done in the same manner as was fixed in `RsModuleIndex` and also follows `RsVisibility.isPublic` logic (that returns true if there is any visibility modifier). If we want to change it, I think it should be another PR

File: src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt (hunk 2 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: Done

File: src/main/kotlin/org/rust/lang/core/stubs/index/RsExternCrateReexportIndex.kt (hunk 1 above)
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/stubs/index/RsExternCrateReexportIndex.kt]: Agree

**Pull request merged**